### PR TITLE
feat(noui): browser-driven skill compilation for encrypted-traffic portals

### PR DESCRIPTION
**Pull request by navjot-adopt**
## Summary

Compile **browser-driven skills** for portals whose traffic can't be HAR-replayed (SPAs that mint per-request encryption / per-session headers — banks like ICICI, HSBCnet, HDFC Life). Instead of replaying HTTP, these skills drive the live page via the harness `call_web_browser` tool and read the rendered DOM. Also hardens the login/keepalive/health compilation the browser path depends on.

> Squash-merged — grouped summary below.

### Browser-driven skill compilation
- New **browser-driven mode** + `--browser-driven` flag wired through `capture_import`. (`ae74587`, `8b7d20d`)
- **Auto-detect unreplayable apps** (encrypted `{data,key}` bodies) and route them to browser mode. (`f7d48ce`)
- Skills navigate by **in-app `click_by_text`, never a full-page reload** (refresh-sensitive portals expire the session on a reload); capture the **full submenu click-chain** (expand → child). (`f5e25e7`, `4250e70`)

### Keepalive — the SPA/bank reload fix
- Keepalive **style follows skill kind**: `activity` (trusted mouse/scroll, no reload) for browser skills, `goto` (revisit for header freshness) for HAR-replay. `activity` is the default nudge. (`ef92a2d`, `bad4da4`)
- Intervals: goto = 120s; activity respects Tabby's 60s floor. (`6a99a11`, `2203aed`)
- Final hardening: one browser decision drives both keepalive style AND skill kind (no browser-skill-with-goto-keepalive); strip `;jsessionid=` matrix params from keepalive/health URLs; decouple `browser_policy.downloads` from keepalive style. (`343c146`)

### Keepalive / health-check URL correctness
- Never replay one-time tokens in keepalive/health URLs; emit no `goto` keepalive when the landing URL can't be replayed; treat session-expiry landings as auth redirects; infer the landing page and emit a health check that can actually FAIL (not a `dom_check` that always passes). (`60158fc`, `2fe942d`, `82cd2f1`, `2063300`)

### Downloads
- Enable `browser_policy.downloads` for browser skills so in-page exports (statement PDFs) can be captured. (`bedb725`)

### Compile-time guardrails (fail loud, not silent)
- Reject skills that silently cannot authenticate, a `--profile-slug` Tabby cannot resolve, or a manifest whose routing contradicts its operations; don't re-prompt a human who already confirmed login. (`604b063`, `68a2272`, `35e56a6`, `4468d8e`)

## Impact / follow-ups
- **Does not affect already-deployed skills** — this changes how NEW skills compile.
- The guardrails now **reject at compile time** skills that previously compiled "successfully but broken" (intended; a behavior change for skill authors).
- Known follow-ups not in this PR: hash-router SPAs (single-URL `#/route` apps like HSBCnet) still can't have their nav derived (`_page_key` collapses on the fragment); text-less icon buttons need `click_element` CSS selectors rather than `click_by_text`.

### CODE REVIEW

File: skills/noui/noui_core/compile/login_assets.py [hunk 1]
```diff
@@ -134,6 +134,64 @@ def _build_selector(ev: dict) -> str:
 # ---------------------------------------------------------------------------
 
 
+# Query params whose value is minted per session/request. A URL carrying one is
+# not replayable: the value died with the recording.
+_VOLATILE_QUERY_PARAM = re.compile(
+    r"(token|jsessionid|sessionid|sid|nonce|ticket|otp|csrf|xsrf|auth|signature|sig|"
+    r"timestamp|_ts|expires)",
+    re.I,
+)
```
[the remark this answers is not in the document]
Comment (navjot-adopt) [reply skills/noui/noui_core/compile/login_assets.py]: Intentional fail-safe. `_VOLATILE_QUERY_PARAM` uses broad substring matching so we never *replay* a one-time token; over-flagging a rare param only skips a `goto` keepalive (which browser skills don't want anyway) — the safe direction. Two notes: the `userid` example doesn't actually match (no `sid` substring in u-s-e-r-i-d), and a `\b` anchor would *miss* real cases like `UX_TOKEN` (no word boundary at `_token`). So the broad match stays.

File: skills/noui/noui_core/compile/browser_skill.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+
+For those apps the workable path is to drive the PAGE: navigate to where the
+data renders, then read the DOM the page already fetched and decrypted. Tabby
+exposes this via `/execute/browser`, and the harness surfaces it to the model as
+the `call_web_browser` tool. This module compiles a recording into a skill that
+uses that tool instead of `call_web_api`.
+
+The "operations" of a browser skill are the readable data PAGES observed in the
+recording — each compiled as: navigate to the page, then `get_page_summary`.
+There is no request replay, so no encryption or frozen-header problem.
```
[the remark this answers is not in the document]
Comment (navjot-adopt) [reply skills/noui/noui_core/compile/browser_skill.py]: Fixed in f677c22 — the docstring now says the compiler reaches each page via the recorded in-app `click_by_text` chain, never a full-page `navigate`/goto (a reload expires the SPA session).

File: skills/noui/references/pillar-2-compile.md [hunk 3]
```diff
@@ -19,6 +19,15 @@ Outputs go under `NOUI_WORKBENCH_DIR` (default `skills/noui/workbench/`): `mcp_s
 
 The vendored `noui_runtime/execute.py` (from `noui_core.activate.execute_adapter`) is what makes a generated asset self-contained and Tabby-`/execute`-backed.
 
+### Browser-driven skills (`--browser-driven`) — for apps replay can't reach
+Some single-page apps encrypt **every request body in the page's JavaScript** with a per-session key fetched at load time (and/or stamp each request with rotating per-request headers). A recorded request is then unreplayable: the body is an opaque `{data, key}` blob only the live page can produce, and the headers die with the recording. ICICI net-banking is the canonical case — a normal HAR-replay skill compiles dozens of operations that **all 403 at run time**, and even a perfect capture of the target endpoint is dead on arrival.
+
+For those apps, compile a **browser-driven** skill: it drives the page (`navigate` → `get_page_summary`) via the harness `call_web_browser` tool and reads the DOM the page already fetched and decrypted, so the encryption is irrelevant. Its "operations" are the readable data pages from the recording, not replayed API calls. Requires a bound `--profile-slug` (it drives an authenticated session).
```
[the remark this answers is not in the document]
Comment (navjot-adopt) [reply skills/noui/references/pillar-2-compile.md]: Fixed in f677c22 — the doc now reads `click_by_text` → `get_page_summary` (never a full-page navigate/reload).

File: skills/noui/scripts/compile_workflow.py [hunk 4]
```diff
@@ -25,6 +25,16 @@ def main() -> int:
...
     p.add_argument("--profile-slug", dest="profile_slug", default="")
+    p.add_argument(
+        "--allow-unbound-profile",
+        dest="allow_unbound_profile",
+        action="store_true",
+        help=(
+            "Emit the skill even when session-authenticated operations have no Tabby "
+            "profile bound. Such a skill cannot authenticate at run time — only use "
+            "this when the binding is applied later."
+        ),
```
[the remark this answers is not in the document]
Comment (navjot-adopt) [reply skills/noui/scripts/compile_workflow.py:36]: `--allow-unbound-profile` is an intentional escape hatch for deferred binding, and its help text already warns the skill *cannot authenticate at run time*. Rejecting it in harness mode is a reasonable future enhancement but out of scope for this PR.

**Comment by navjot-adopt** [comment]
## Deep review — findings before stage

Full-diff review vs `dev`; findings reproduced by running the compiler with the repo venv or by reading the consuming code in `tabby`. Full suite passes (646) — **none of these are caught by tests**. Nothing changed yet.

### 🔴 Blocking

**1. `login_assets.py:933` — cross-origin SSO logins now pick the IdP page as `post_login_url`.**
`_is_redirect_hop` returns False for a cross-netloc hop (`:320`), so an SSO bounce stays in `stable_urls` and the new "first same-origin run" walk stops at index 0 = the IdP. Reproduced with `app.contoso.com/login → idp.okta.com/signin → app.contoso.com/home`: emits `wait_for_url "https://idp.okta.com/**"` and a `url_check` on the IdP whose own `auth_redirect_pattern` contains `/signin` ⇒ **permanent AUTH_FAIL every keepalive cycle**. The old `stable_urls[-1]` picked correctly.
*Fix:* skip leading `stable_urls` entries that are login-flow URLs or whose origin ≠ the login origin, then apply the same-origin walk.

**2. `login_assets.py:1105` — `auth_redirect_pattern` alternatives are unanchored substrings tested against the whole URL, so healthy landing pages get classified AUTH_FAIL.**
The worker does `new RegExp(pattern,'i').test(page.url())` on the **full** URL and runs it *before* the HTTP probe. Verified matches: `https://auth.example.com/dashboard` (`//auth`), `https://sso.corp.com/home`, `https://app.test/authorize-payments`, `https://app.test/author/5`, `.../corp/AuthenticationController` (`/Auth`), and — via the `re.escape(_login_path)` entry — `https://x.test/app/dashboard` when login lives at `/app`. Tabby's own `looksLikeAuthUrl` deliberately uses `\b…\b` "to avoid false positives like /authorization-settings"; this is looser. Compounding: the worker suppresses the `activity` nudge whenever the last health wasn't PASS, so a false AUTH_FAIL also **kills the keepalive** — the very failure this PR fixes.
*Fix:* anchor each alternative (`(^|/)login(/|$|\?)`), drop bare `/auth` in favour of word boundaries, and exclude any pattern that matches `post_login_url` itself before emitting.

**3. `skill_generator.py:115` — `--auth-type api-key` with no `--profile-slug` (the documented correct combination) now fails to compile.**
Reproduced: a HAR with `authorization: Bearer sk_…`, `auth_type="api-key"`, `profile_slug=""` raises "…session-authenticated … but no Tabby profile was bound". `har_to_tool_defs` fills `auth_headers` regardless of `declared_strategy` and the guard never consults it; `pillar-2-compile.md` defines api-key as "no login recorded" → no profile by design. The second guard at `:365` already excludes this case; the first doesn't.
*Fix:* skip the guard when `declared_strategy == "static_secret_header"`.

**4. `browser_skill.py:188` — multi-hop pages emit an unreachable recipe (only the last hop's click).**
`nav` is the chain recorded on `from_url`, never accumulated back to the landing page. Reproduced `home → accounts → statements`: `read_statements.steps == [click_by_text "Statements", get_page_summary]`, but at run time the session lands on `/home` where "Statements" doesn't exist.
*Fix:* prefix each page's steps with the recursive nav chain of its `from_url`.

**5. `browser_skill.py:188` — "no click found" is indistinguishable from "landing page", so an operation silently reads the wrong page.**
`_nav_clicks_for(...) or None` → a `home → accounts` hop whose driving click had empty `text_content` (icon/SVG button — very common in bank navs) yields `nav: null` ⇒ steps are just `get_page_summary`, returning the **landing page's** DOM while the op claims to read `/accounts`. Confidently wrong data is worse than a failure.
*Fix:* only treat `nav is None` as landing when `_is_login_flow_url(from_url) or not from_url`; otherwise emit a review item / drop the page.

**6. `browser_skill.py:51` — hash-router SPAs collapse to a single page, silently.**
`_page_key` = origin + path, dropping the fragment. Reproduced `/app#/home`, `/app#/accounts`, `/app#/statements` → one page, `nav: null`. `generate_browser_skill` only raises when the list is *empty*, so this compiles and installs a skill that can read only the landing screen. Playwright does fire `framenavigated` on hash changes, so the events are recorded and then discarded. *(This is the HSBCnet shape.)*
*Fix:* include a `#/…` fragment in `_page_key`/`_slug_from_path` and route it via `click_by_text`.

### 🟡 Non-blocking

7. `capture_import.py:144` — `--keepalive` now silently decides the **skill kind**: `_resolve_keepalive_style` returns an explicit `--keepalive` before consulting `args.browser_driven`, and `is_browser = style == "activity"` feeds `browser_driven=` with `auto_detect_browser=False`. So `--browser-driven --keepalive goto` compiles a **HAR-replay** skill (explicit flag ignored), `--keepalive activity` forces browser mode on a replayable app, and `--keepalive goto` on an ICICI-shaped capture disables auto-detection entirely with no `browser_detection` report.
8. `workflow.py:75` — a 401/403 on `/admin/profiles` is converted to a definite "does not exist" (`get_service_profile_by_slug` swallows `RuntimeError` → `None`), hard-failing a valid compile with `Tabby profile 'x' does not exist`. The code's own comment anticipates exactly this.
9. `login_assets.py:1133` — only *matrix* params are stripped from the health URL; a volatile **query** (ICICI's `?…&UX_TOKEN=<one-time>`) is still probed every 60s, yielding either a falsely-PASS error page or permanent TRANSIENT_FAIL (which also suppresses the nudge, per finding 2).
10. `browser_skill.py:133` — `out[-_NAV_MAX_CLICKS:]` drops the **leading** entry-nav click. Reproduced with `Menu → Banking → Cards → Credit Card`: emits the last 3, so the first click targets a still-hidden element. Three-deep accordions are normal on HSBCnet/ICICI.
11. `browser_skill.py:105` — a `url_event` with no timestamp disables the causality filter while still anchoring on the last click, so the chain can be *post*-navigation clicks (reproduced: a compiled `read_accounts` that clicks "Log out" first). Tabby's own recorder always stamps timestamps, hence non-blocking.
12. `capture_import.py:444` — the `--mode login` path ignores `--keepalive`/`--browser-driven`/downloads, so a login-only recording for a browser portal always registers `goto`/120s/`downloads:False`.
13. `skill_generator.py:365` — `--allow-unbound-profile` doesn't work with `--execution-mode harness` (still raises), so the flag's documented contract is false for the mode this PR targets.
14. `login_assets.py:139` — `_VOLATILE_QUERY_PARAM` is a bare substring alternation: verified True for `?assignee=`, `?author=`, `?ticket=`, `?sidebar=`, `?resident=`, `?design=`. Cheap in `login_assets` (skips a goto) but `browser_skill.py:175` **discards the whole page** on a hit, so a portal whose data screens all carry e.g. `?ticket=` yields zero pages → `ValueError: No readable data page was captured`.
16. `compile_workflow.py` — the docs and `unreplayable.recommendation_message` tell operators to pass `--browser-driven`/`--no-auto-browser`, but this entry point never gained them, and `auto_detect_browser` defaults True — so re-compiling a saved bundle can flip to browser mode with no way to turn it off.

### 🧪 Test gaps
- `_resolve_keepalive_style` and the `is_browser` coupling in `_run_combined` — the core of this PR — have **zero** tests (`test_combined_import.py` untouched).
- No test for cross-origin SSO `post_login_url` selection, for `auth.*`/`sso.*` hosts or `/authorize` against the auth pattern, for a multi-hop nav chain, a hash-router recording, a nav-less non-landing page, `_NAV_MAX_CLICKS` truncation, or a negative case for `_has_volatile_query`.

### Nits
- `_strip_volatile_matrix_params` partitions on `?` first, so it eats a **fragment** when there's no query: `…/app#/report;sid=5` → `…/app#/report`. Partition on `#` first.
- The "Browser mode auto-selected" report is dead on the combined path (`auto_detect_browser=False` ⇒ `browser_detection` never populated).
- Generated `SKILL.md` never mentions `list_downloads`/`get_download`, so the model can't discover the capability the downloads change enables.
- `unreplayable.py:104` — `key_endpoints` is collected across all origins although the docstring says `app_origin` scopes detection.

**Comment by navjot-adopt** [comment]
### Addressed in `1243e27` — all six blocking findings

**1. Cross-origin SSO landing page** — leading sign-in pages are skipped before the same-origin walk, so `app/login → idp/signin → app/home` now infers `app.contoso.com`, not the IdP. Falls back to the old behaviour if that would leave nothing.

**2. `auth_redirect_pattern` anchoring** — each alternative is now `(?<!/)/segment(?:[/?#]|$)`, which kills the whole false-positive class (the `//auth` of an `auth.*` host, `sso.*` hosts, `/authorize-payments`, `/author/5`). The recorded login path matches as a *whole* path only, so a login at `/app` no longer flags `/app/dashboard`. Plus a backstop: any pattern that matches the landing URL itself is dropped before emitting — a self-matching pattern is a permanent AUTH_FAIL, and since the `activity` keepalive only nudges while health is PASS, it would kill the session it watches.

**3. api-key + no profile** — the unbound-profile guard now exempts `static_secret_header`, matching the manifest guard that already did.

**4. Multi-hop nav** — `_resolve_nav_chains` walks the from-page graph back to the landing page, so `landing → accounts → statements` emits `["Accounts", "Statements"]` instead of just the last hop.

**5. Undeterminable nav** — `nav` now distinguishes three states (`None` landing / `[...]` chain / `[]` unrecoverable). A non-landing page whose driving click can't be recovered is **dropped** rather than compiled into a recipe that silently reads the landing DOM.

**6. Hash-router pages** — the `#/route` fragment is part of `_page_key` and the slug, so HSBCnet-shaped SPAs no longer collapse to one page. Bare `#anchor` fragments still aren't routes (they'd split one page into many).

**Verification:** 654 passed (12 new tests, one per finding plus edge cases). `ruff check` and `ruff format --check` both clean under CI's ruff 0.16.1.

One note worth flagging: three existing tests in `test_browser_skill.py` were passing *because* of finding 5 — the ICICI fixture's own comment said "in-app click drove this route change → nav = click 'Credit Cards'" but no click was ever supplied, and the old `or None` turned that into landing semantics. The fixture now carries the click it always described.

**Still open from the review:** non-blocking 7–14 and 16, and the test gaps — notably 7 (`--keepalive` silently deciding skill kind, so `--browser-driven --keepalive goto` compiles a HAR-replay skill) and 14 (`_VOLATILE_QUERY_PARAM` substring over-match discarding whole pages in `browser_skill`).

**Comment by navjot-adopt** [comment]
### Non-blocking findings addressed in `7103189`

**7 (flag precedence).** `--browser-driven` now wins over `--keepalive`. `_run_combined` derives the **skill kind** from `_resolve_keepalive_style`, so checking the explicit `--keepalive` first meant `--browser-driven --keepalive goto` silently compiled a HAR-replay skill, ignoring the flag the caller passed.

**14 (volatile params) — fixed in two halves.** Matching is now on whole param-name **tokens** rather than bare substrings, so ordinary routing params stop firing (`?assignee=`/`?design=` for "sig", `?author=` for "auth", `?sidebar=`/`?resident=` for "sid"). And the structural half, which was the real bug: `browser_skill` no longer **discards** a page whose recorded URL has a volatile query. A one-time token makes a URL unreplayable *by navigation* — but a browser skill never navigates, it replays the click chain, and `_page_key` ignores the query anyway. A portal whose data screens all carry e.g. `?ticket=` compiled to zero readable pages.

**10 (nav truncation).** The cap keeps the **first** click plus the most recent ones. A trailing slice dropped the entry gesture: `Menu → Banking → Cards → Credit Card` became the last three, so the first click targeted an element still hidden behind the un-opened hamburger.

**11 (causality).** A `url_event` with no timestamp falls back to the single navigating click — without one the causality filter cannot run, so the "clicks before the nav" set could actually be clicks made *after* it (a recorded "Log out" compiled into the read recipe).

**9 (volatile health URL).** When the query is volatile the probe falls back to the origin root. The query often *is* the routing so it cannot be stripped, and probing a dead token every 60s yields either a falsely-PASS error page or a permanent TRANSIENT_FAIL — which also suppresses the activity nudge.

**8 (forbidden ≠ absent).** `_profile_slug_resolves` probes `/admin/profiles` before trusting the lookup. `get_service_profile_by_slug` swallows `RuntimeError` — including a 401/403 from that Editor+-gated endpoint — and returns `None`, indistinguishable from "no such profile", so an agent token that can read templates but not profiles hard-failed a valid compile with *"Tabby profile 'x' does not exist"*.

**12 (login-mode flags).** The `--mode login` path honours `--keepalive`/`--browser-driven` and downloads; it previously always registered goto/120s with downloads off regardless of what was passed.

**Deliberately not changed — 13 and 16** (`--allow-unbound-profile` semantics under `--execution-mode harness`, and adding the browser flags to `compile_workflow.py`). Both are CLI-surface changes that want their own PR rather than riding along here.

**Verification:** 655 passed; `ruff check` and `ruff format --check` clean under CI's ruff 0.16.1.

**Review by rahulbh1510 — APPROVED** [review]
